nemo/collections/asr/parts/utils/diarization_utils.py

```diff
@@ -748,84 +426,288 @@ def write_json_and_transcript(
+            word_pos = word_ts_stt_end[1]
+        elif self.params['word_ts_anchor_pos'] == 'mid':
+            word_pos = (word_ts_stt_end[0] + word_ts_stt_end[1]) / 2
+        else:
+            logging.info(
+                f"word_ts_anchor_pos: {self.params['word_ts_anchor']} is not a supported option. Using the default 'start' option."
+            )
+            word_pos = word_ts_stt_end[0]
+
+        word_pos = word_pos + self.word_ts_anchor_offset
+        return word_pos
+
+    @experimental
+    def realign_words_with_lm(self, word_dict_seq_list: List[Dict[str, float]]):
+        """
+        Realign the mapping between speaker labels and words using a language model.
+        The realigning process calculates the probability of the around the words at the boundary between
+        two hypothetical sentences spoken by different speakers.
+
+        <Example> k-th word: "but"
+            hyp_former:
+                since i think like tuesday </s> <s>  but he's coming back to albuquerque
+            hyp_latter:
+                since i think like tuesday but </s> <s>  he's coming back to albuquerque
+
+        The joint probabilities of words in the sentence are computed for these two hypotheses. In addition,
+        logprob_diff_threshold parameter is used for reducing the false positive realigning.
+        """
+        word_seq_len = len(word_dict_seq_list)
+        hyp_w_dict_list, spk_list = [], []
+        for k, line_dict in enumerate(word_dict_seq_list):
+            word, spk_label = line_dict['word'], line_dict['speaker_label']
+            hyp_w_dict_list.append(word)
+            spk_list.append(spk_label)
+
+        realigned_list = []
+        org_spk_list = copy.deepcopy(spk_list)
+        for k, line_dict in enumerate(word_dict_seq_list):
+            if self.N_range[0] < k < (word_seq_len - self.N_range[0]) and (
+                spk_list[k] != org_spk_list[k + 1] or spk_list[k] != org_spk_list[k - 1]
+            ):
+                N1, N2 = self.get_realignment_ranges(k, word_seq_len)
+                hyp_former = self.realigning_lm.log_s(
+                    ' '.join(hyp_w_dict_list[k - N1 : k] + self.stt_end_tokens + hyp_w_dict_list[k : k + N2])
+                )
+                hyp_latter = self.realigning_lm.log_s(
+                    ' '.join(hyp_w_dict_list[k - N1 : k + 1] + self.stt_end_tokens + hyp_w_dict_list[k + 1 : k + N2])
+                )
+                log_p = [hyp_former, hyp_latter]
+                p_order = np.argsort(log_p)[::-1]
+                if log_p[p_order[0]] > log_p[p_order[1]] + self.realigning_lm_params['logprob_diff_threshold']:
+                    if p_order[0] == 0:
+                        spk_list[k] = org_spk_list[k + 1]
+                line_dict['speaker_label'] = spk_list[k]
+            realigned_list.append(line_dict)
+        return realigned_list
+
+    def get_alignment_errors(self, ctm_content, hyp_w_dict_list, mapping_dict):
```

Comment:
Maybe we can move and gather evaluation codes, i.e. alignement error, der,  in another utils files to slim diarization utils later

Reply:
If that is the plan, bare-diarization is better to be arranged to one place. It is a thing to be discussed.